apps/prairielearn/python/zygote.py

```diff
@@ -84,6 +87,33 @@
 prairielearn.get_unit_registry()
 
 
+# We want to conditionally allow/block importing specific modules, namely `rpy2`.
+# This custom importer will allow us to do so, and throw a custom error message.
+#
+# While this won't prevent anything more complex than an `import` statement, it
+# will make it clear to the user that they're not allowed to use `rpy2`. If they
+# try to bypass the block, it's up to them to deal with the consequences.
+class ForbidModuleMetaPathFinder(MetaPathFinder):
+    def __init__(self):
+        self.forbidden_modules: set[str] = set()
```

Comment:
```suggestion
        self.__forbidden_modules: set[str] = set()
```
nit: Only if you REALLY don't want client code to alter this set directly: https://realpython.com/python-double-underscore/#double-leading-underscore-in-classes-pythons-name-mangling

Reply:
Are you thinking that a particularly clever client might reach into `sys.meta_path` and change things? Even with the leading underscores, they could get our path finder into whatever state they want with its public methods. And frankly, I don't care if someone tries that, because they deserve to have their question broken when we inevitable remove R entirely 😅

Reply:
It's less a concern and more just over-engineering things in my head 😂😂